oeps/oep-0013.rst

```diff
@@ -0,0 +1,672 @@
+* **404** - Not Found
+* **415** - Unsupported Media Type (used for PATCH if implementation is "merge
+            patch" algorithm, and caller did not specify
+            "application/merge-patch+json" content type)
+* **500** - Internal Server Error
+
+**Prevent information leakage**
+
+    **Must**: Use 404 instead of 403 when the actual existence would be leaking
+    information that we don't want
+
+**Error description**
+
+    **Must**: Be verbose and use plain language descriptions.
+
+    **Must**: Use i18n for user facing messages ("user_message" in example
+    below)
+
+    **Nice**: Add as many hints as your API team can think of about what's
+    causing an error.
+
+    **Nice**: Add a link in your description to more information, like Twilio
+    does.
+
+**Error format**
+
+    - ``developer_message`` (no i18n)
+    - ``user_message`` (optional, but if provided, i18n as the expectation is
+      that this may surface in a UI)
+    - ``field_errors`` (if applicable)
+    - ``error_code`` (for example: see LMS ``has_access`` interface design)
+        *Include description in your REST API Docs.*
+
+    ::
+
+        {
+            "error_code": "course_not_started"  # a short string that the client can rely upon for handling different errors
+            "developer_message": "Verbose, plain language description of the problem for the app developer with hints about how to fix it.",
+            "user_message": "Pass this message on to the app user if needed.",
+            "field_errors": {
+                "foo": {
+                    "developer_message": "",
+                    "user_message": ""
+                }
+            }
+        }
+
+
+Versioning
+----------
+
+**Location**
+
+- Put the version in the URL so the client can see it easily when handling the
+  response logic.
+
+- Exception: Put it in the header only if it doesn't change the response
+  handling logic, like the OAuth endpoint.
+
+**Numbering**
+
+- *Must*: Versions are major only ex. ``/v0/`` ``/v1/`` ``/v2/``
+
+- *Must*: Major version should only be updated when changing the existing
+  contract in a way that breaks backwards compatibility.  The previous major
+  version should continue to be supported for old clients according to the
+  service's deprecation policy.
+
+- *Must*: Additive changes to the contract of the API, or the results, should
+  not bump the major version, unless those changes are non-optional, and break
+  existing contracts with the API.
+
+Pagination
+----------
+
+We use the conventions established by `DRF
+<http://www.django-rest-framework.org/api-guide/pagination/#pagenumberpagination>`_,
+with some extra fields:
+
+- ``next``: A URL to the next page, or null.
+- ``previous``: A URL to the previous page, or null.
+- ``count``: The total number of items.
+- ``num_pages``: The total number of pages.
+- ``results``: List of results.
+
+The query parameters are as follows:
+
+- ``page`` - the page of results (zero indexed)
```

Comment:
Why zero-indexed? I can't think of any of our own APIs that implement this. If we want to see this, and the other conventions (e.g. num_pages), implemented we should make it easy to do so by creating a pagination class in https://github.com/edx/edx-drf-extensions.